### PR TITLE
Inaccurate Metadata Fixed

### PR DESCRIPTION
**Pull request by sadrasabouri**
#### Reference Issues/PRs
#165 

#### What does this implement/fix? Explain your changes.
I decided to add functions in `utils.py` so that the following parameters are justified in the generated file metadata:

- `max-weight`: maximum weight of edges in the graph, determined by `weight_dic`
- `min-weight`: minimum weight of edges in the graph, determined by `weight_dic`
- `self-loop`: whether the graph contains self-loop or not, determined by `edge_dic` if it has the same vertex in the destination list for a vertex
- `multigraph`: if the graph is multigraph, determined by `edge_dic` if it has nonunique members in the vector destination list.

The following metadata removed because it was the same with `edge_number` in graphs:

- `max-edge`
- `min-edge`

### CODE REVIEW

File: pyrgg/engines/pyrgg.py [hunk 1]
```diff
@@ -236,15 +238,13 @@ def gen_using(
...
+            "max_weight": get_max_weight(weight_dic),
             "vertices_number": input_dict['vertices'],
-            "min_edge": input_dict['min_edge'],
-            "max_edge": input_dict['max_edge'],
-            "sign": input_dict['sign'],
+            "sign": is_signed(weight_dic),
             "direct": input_dict['direct'],
-            "self_loop": input_dict['self_loop'],
-            "multigraph": input_dict['multigraph'],
+            "self_loop": has_self_loop(edge_dic),
```
Comment (sepandhaghighi) [inline pyrgg/engines/pyrgg.py:246]: I recommend removing the `self_loop` from the metadata.

File: pyrgg/engines/pyrgg.py [hunk 2]
```diff
@@ -236,15 +238,13 @@ def gen_using(
...
+            "multigraph": is_multigraph(edge_dic),
```
Comment (sepandhaghighi) [inline pyrgg/engines/pyrgg.py:247]: I recommend removing the `multigraph` from the metadata.

File: pyrgg/engines/pyrgg.py [hunk 3]
```diff
@@ -236,15 +238,13 @@ def gen_using(
...
             "file_name": file_name,
-            "min_weight": input_dict['min_weight'],
-            "max_weight": input_dict['max_weight'],
+            "min_weight": get_min_weight(weight_dic),
+            "max_weight": get_max_weight(weight_dic),
             "vertices_number": input_dict['vertices'],
-            "min_edge": input_dict['min_edge'],
-            "max_edge": input_dict['max_edge'],
-            "sign": input_dict['sign'],
+            "sign": is_signed(weight_dic),
```
Comment (sepandhaghighi) [inline pyrgg/engines/pyrgg.py:244]: I recommend removing the `sign` from the metadata.

File: pyrgg/utils.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    :param edge_dic: edge dictionary
+    :type edge_dic: dict
+    :param weighted: weighted flag
+    :type weighted: bool
+    :return: maximum weight value
+    """
+    return max([abs(w) for weights in weight_dic.values() for w in weights])
+
+
+def is_signed(weight_dic):
```
Comment (sepandhaghighi) [inline pyrgg/utils.py:31]: You may retain the `has_self_loop`, `is_signed`, and `is_multigraph` functions for potential future use, but remember to include `#pragma: no cover` at their first line.

File: pyrgg/utils.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+"""Pyrgg utility module."""
+
+
+def get_min_weight(weight_dic):
```
Comment (sepandhaghighi) [inline pyrgg/utils.py:5]: I suggest combining the `get_min_weight` and `get_max_weight` functions to improve efficiency.

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@sadrasabouri, thank you for your efforts 💯 
Please review my comments

**Comment by sadrasabouri** [comment]
@sepandhaghighi I guess we shouldn't remove metadata, at least for `gml`. Take a look at failing tests which raised when importing a multi-graph without metadata to networkx.
```
networkx.exception.NetworkXError: edge #17 (12--25) is duplicated
```

**Comment by sepandhaghighi** [comment]
> @sepandhaghighi I guess we shouldn't remove metadata, at least for `gml`. Take a look at failing tests which raised when importing a multi-graph without metadata to networkx.
> 
> ```
> networkx.exception.NetworkXError: edge #17 (12--25) is duplicated
> ```

@sadrasabouri 

You are right, the `gml` format supports `multigraph`. Therefore, we need to revert the `multigraph` meta data. I suggest adding it to other formats as well.

File: pyrgg/graph_gen.py [hunk 6]
```diff
@@ -90,38 +85,27 @@ def _write_data_to_json(buf, edge_dic, weight_dic):
 
 def _write_properties_to_json(
         buf,
-        min_weight,
```
Comment (sepandhaghighi) [inline pyrgg/graph_gen.py:93]: I have a question. Why did you remove `min_weight` and `max_weight` from the JSON format?

File: test/engines/pyrgg_test.py [hunk 7]
```diff
@@ -561,29 +539,17 @@
 76
 >>> testfile_3['graph']['edges'][1]['weight']
 15
->>> testfile_3['properties']['signed']
-False
 >>> testfile_3['properties']['directed']
 True
->>> testfile_3['properties']['self_loop']
-True
->>> testfile_3['properties']['multigraph']
```
Comment (sepandhaghighi) [inline test/engines/pyrgg_test.py:570]: Please revert **multigraph** tests

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: pyrgg/graph_gen.py (hunk 6 above)
Comment (sadrasabouri) [reply pyrgg/graph_gen.py:93]: Since they were only used for calculating for `weighted`, now it's passed to this function directly. Therefore we don't need them anymore. The reason was that `is_weighted` could be used when calling `gen_function` inside `gen_using`. Just for the case of consistency.

File: test/engines/pyrgg_test.py (hunk 7 above)
Comment (sadrasabouri) [reply test/engines/pyrgg_test.py:570]: fixed in [25f71d4](https://github.com/sepandhaghighi/pyrgg/pull/173/commits/25f71d4b3f5fb271376342350900f6a59fa5fc50).

**Comment by sadrasabouri** [comment]
Please re-review @sepandhaghighi 
Also I guess we may need to also update the logger with the correct metadata, too. BTW, it's out of this PR's scope.

File: test/graph_gen_test.py [hunk 8]
```diff
@@ -135,14 +105,8 @@
 76
 >>> testfile_3_p['graph']['edges'][1]['weight']
 15
->>> testfile_3_p['properties']['signed']
-False
 >>> testfile_3_p['properties']['directed']
 True
->>> testfile_3_p['properties']['self_loop']
-True
->>> testfile_3_p['properties']['multigraph']
```
Comment (sepandhaghighi) [inline test/graph_gen_test.py:144]: Please revert all `multigraph` tests!

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: test/graph_gen_test.py (hunk 8 above)
Comment (sadrasabouri) [reply test/graph_gen_test.py:144]: Sorry for that; it was fixed in the new commit.

**Comment by sadrasabouri** [comment]
Sorry because of the mistake, please re-review @sepandhaghighi when you have time.

File: pyrgg/functions.py [hunk 9]
```diff
@@ -28,6 +28,53 @@ def is_weighted(max_weight, min_weight, signed):
     return True
 
 
+def get_min_max_weight(weight_dic):
+    """
+    Get minimum and maximum weight values.
```
Comment (sepandhaghighi) [inline pyrgg/functions.py:33]: This function's docstring is incorrect.

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: pyrgg/functions.py (hunk 9 above)
Comment (sadrasabouri) [reply pyrgg/functions.py:33]: Oh shoot. Yes, that's again residual from the last commits. fixed in [cc0f41f](https://github.com/sepandhaghighi/pyrgg/pull/173/commits/cc0f41f339eb1cfd0dfdcd16007928967c2a8c88).

**Review by sepandhaghighi — APPROVED** [review]
🚀
